### PR TITLE
Refacto odo dev

### LINKED ISSUE #6159: Make pkg/dev platform agnostic
/kind feature

The pkg/dev package is for the moment designed for working on Kubernetes platform only.

## Which functionality do you think we should add?

Make the package platform agnostic.

## Why is this needed?

So we can reuse it for other platforms (Podman)

### PR DESCRIPTION
**Pull request by feloy**
**What type of PR is this:**

/kind cleanup

**What does this PR do / why we need it:**

**Which issue(s) this PR fixes:**

Part of #5865 
Fixes #6159 

**PR acceptance criteria:**

- [ ] Unit test 

- [ ] Integration test 

- [ ] Documentation 

**How to test changes / Special notes to the reviewer:**

### CODE REVIEW

**Review by valaparthvi — APPROVED** [review]
Nice refactor!

**Comment by dharmit** [comment]
> Fixes #6159

How do we know that the PR fixes this issue? Have we tried this with podman?

**Comment by feloy** [comment]
> > Fixes #6159
> 
> How do we know that the PR fixes this issue? Have we tried this with podman?

If you look at the code into pkg/odo/cli/dev, you should not see any code related to Kubernetes (1).

Also, into the interface definition in pkg/dev, no parameter of the methods should be related to Kubernetes.

(1) There is still a problem about calling:
```
o.clientset.KubernetesClient.SetNamespace(o.GetProject())
```
but this is a generic problem how we are managing the context, not specific to the `odo dev` command.

File: pkg/dev/dev.go [hunk 1]
```diff
@@ -2,20 +2,24 @@ package dev
...
+	"fmt"
 	"io"
 
 	"github.com/redhat-developer/odo/pkg/binding"
 	"github.com/redhat-developer/odo/pkg/envinfo"
 	"github.com/redhat-developer/odo/pkg/kclient"
 	"github.com/redhat-developer/odo/pkg/portForward"
 	"github.com/redhat-developer/odo/pkg/preference"
-	"github.com/redhat-developer/odo/pkg/testingutil/filesystem"
+	filesystem "github.com/redhat-developer/odo/pkg/testingutil/filesystem"
```
Comment (rm3l) [inline pkg/dev/dev.go:13]: nit: is this import alias needed? I find it a bit redundant.
https://github.com/redhat-developer/odo/wiki/Dev:-Coding-Conventions#import-aliasing

File: pkg/dev/interface.go [hunk 2]
```diff
@@ -17,44 +16,29 @@ type Client interface {
...
-
-	// Watch watches for any changes to the files under path while ignoring the files/directories in ignorePaths.
+	//
+	// It then watches for any changes to the files under path while ignoring the files/directories in ignorePaths.
 	// It logs messages to out and uses the Handler h to perform push operation when anything changes in path.
 	// It uses devfileObj to notify user to restart odo dev if they change endpoint information in the devfile.
 	// If debug is true, the debug command will be started after a sync, or the run command by default.
 	// If buildCommand is set, this will look up the specified build command in the Devfile. Otherwise, it uses the default one.
 	// If runCommand is set, this will look up the specified run command in the Devfile and execute it. Otherwise, it uses the default one.
 	// componentStatus is the status returned from the call to the Start Method
```
Comment (rm3l) [inline pkg/dev/interface.go:26]: Can you remove the doc about `componentStatus` since it is not a parameter of `Start`?

**Review by rm3l — COMMENTED** [review]
LGTM - thanks for this!

Added a few nitpicky comments.

File: pkg/dev/interface.go [hunk 3]
```diff
@@ -17,44 +16,29 @@ type Client interface {
...
 		buildCommand string,
 		runCommand string,
-		variables map[string]string,
 		randomPorts bool,
 		watchFiles bool,
+		variables map[string]string,
+		out io.Writer,
 		errOut io.Writer,
-		componentStatus watch.ComponentStatus,
 	) error
```
Comment (dharmit) [inline pkg/dev/interface.go:42]: IMO, we should separate optional parameters into a struct. This is too large a number of arguments for the function, no?

File: pkg/dev/dev.go [hunk 4]
```diff
@@ -25,6 +29,7 @@ type DevClient struct {
 	portForwardClient portForward.Client
 	watchClient       watch.Client
 	bindingClient     binding.Client
+	filesystem        filesystem.Filesystem
 }
 
 var _ Client = (*DevClient)(nil)
```
Comment (dharmit) [inline pkg/dev/dev.go:35]: Could we ensure here that the `Handler` interface is implemented since it's no longer on the CLI but on the business layer?

**Review by dharmit — CHANGES_REQUESTED** [review]

File: pkg/dev/interface.go (hunk 2 above)
Comment (feloy) [reply pkg/dev/interface.go:26]: I have also rewritten the doc

File: pkg/dev/dev.go (hunk 4 above)
Comment (feloy) [reply pkg/dev/dev.go:35]: This Handler is not necessary anymore, removing it.

**Comment by dharmit** [comment]
Thanks for the discussion and the work @feloy.

/lgtm

**Comment by feloy** [comment]
/override ci/prow/v4.11-integration-e2e
Tests pass on IBM Cloud

**Pull request merged**